app/src/main/java/emu/skyline/preference/IntegerListPreference.kt

```diff
@@ -0,0 +1,329 @@
+import android.util.AttributeSet
+import android.content.Context
+import android.content.res.Resources
+import android.content.res.TypedArray
+import android.content.DialogInterface
+import android.annotation.SuppressLint
+import androidx.annotation.ArrayRes
+import androidx.appcompat.app.AlertDialog
+import androidx.core.content.res.TypedArrayUtils
+import androidx.preference.R
+import androidx.preference.DialogPreference
+import androidx.preference.PreferenceDialogFragmentCompat
+import emu.skyline.R as sR
+import emu.skyline.di.getSettings
+
+/**
+ * A Preference that displays a list of entries as a dialog.
+ * This preference saves an integer value instead of a string one.
+ * @see androidx.preference.ListPreference
+ */
+@SuppressLint("RestrictedApi", "ResourceType")
+class IntegerListPreference @JvmOverloads constructor(
+    context : Context,
+    attrs : AttributeSet? = null,
+    defStyleAttr : Int = TypedArrayUtils.getAttr(
+        context, R.attr.dialogPreferenceStyle,
+        R.attr.dialogPreferenceStyle
+    ),
+    defStyleRes : Int = 0
+) : DialogPreference(context, attrs, defStyleAttr, defStyleRes) {
+    /**
+     * The list of entries to be shown in the list in subsequent dialogs.
+     */
+    var entries : Array<CharSequence>?
+
+    /**
+     * The array to find the value to save for a preference when an entry from entries is
+     * selected. If a user clicks on the second item in entries, the second item in this array
+     * will be saved to the preference.
+     */
+    var entryValues : IntArray?
+
+    private var value : Int? = null
+        set(value) {
+            // Always persist/notify the first time.
+            val changed = field != value
+            if (changed || !isValueSet) {
+                field = value
+                isValueSet = true
+                value?.let { persistInt(it) }
+                if (changed) {
+                    notifyChanged()
+                }
+            }
+        }
+
+    private var isValueSet = false
+    val refreshRequired : Boolean
+
+    init {
+        val res : Resources = context.resources
+
+        val a = context.obtainStyledAttributes(
+            attrs, sR.styleable.IntegerListPreference, defStyleAttr, defStyleRes
+        )
+
+        entries = TypedArrayUtils.getTextArray(
+            a, sR.styleable.IntegerListPreference_entries,
+            sR.styleable.IntegerListPreference_android_entries
+        )
+
+        val entryValuesId = TypedArrayUtils.getResourceId(
+            a, sR.styleable.IntegerListPreference_android_entryValues,
+            sR.styleable.IntegerListPreference_android_entryValues, 0
+        )
+        entryValues = if (entryValuesId != 0) res.getIntArray(entryValuesId) else null
+
+        if (TypedArrayUtils.getBoolean(
+                a, sR.styleable.IntegerListPreference_useSimpleSummaryProvider,
+                sR.styleable.IntegerListPreference_useSimpleSummaryProvider, false
+            )
+        ) {
```

Comment:
This is borderline unreadable, don't use this much spacing and especially don't do it in an `if` statement